astropy/coordinates/name_resolve.py

```diff
@@ -152,12 +167,58 @@ def get_icrs_coordinates(name, parse=False):
             fmt_url = fmt_url.format(name=urllib.parse.quote(name), db=db)
             urls.append(fmt_url)
 
+    if cache:
+        try:
+            dldir, urlmapfn = _get_download_cache_locs()
+        except OSError as e:
+            msg = 'Remote data cache could not be accessed due to '
+            estr = '' if len(e.args) < 1 else (': ' + str(e))
+            warn(data.CacheMissingWarning(msg + e.__class__.__name__ + estr))
+            cache = False
+
     exceptions = []
     for url in urls:
+
+        if cache and not overwrite:
+            with shelve.open(urlmapfn) as url2hash:
```

Comment:
It is potentially bug-prone not to lock on read. Maybe it's fine. But better to leave that kind of worry inside astropy.utils.data.